Coherence test when a generic type param has a default value from an associated type

ohadravid opened the pull request:
A followup on #61400.
Before `re_rebalance_coherence`, this fails to compile (even though it should be accepted).
`re_rebalance_coherence` had no direct test for this, and I wanted to (a) make sure it doesn't regress in the future and (b) get it on record that this is actually the intended behavior.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

ohadravid commented:
r? @jonas-schievink

jonas-schievink commented:
r? @nikomatsakis

Dylan-DPC commented:
ping from triage @nikomatsakis @jonas-schievink any updates on this?

nikomatsakis commented on src/test/run-pass/coherence/re-rebalance-coherence-default-generic-associated-type.rs:
```diff
@@ -0,0 +1,22 @@
+    type Owned = String;
+}
+
+impl PartialEq<MyString> for LibCow<MyString> {
```
This test doesn't really *check* what the default works out to be, right? Should we do that?

nikomatsakis commented on src/test/run-pass/coherence/re-rebalance-coherence-default-generic-associated-type.rs:
```diff
@@ -0,0 +1,22 @@
+#![allow(dead_code)]
+#![feature(re_rebalance_coherence)]
+
+// run-pass
```
Can we leave a comment elaborating what this test is trying to test?

ohadravid commented:
Hi @nikomatsakis!
Thanks for looking at this. I added a comment about what we are testing here.
I also added a check that we really get the expected type.

nikomatsakis commented:
@bors r+ rollup

nikomatsakis commented:
@bors r+

lqd commented:
@bors r=nikomatsakis rollup

bors commented:
:pushpin: Commit 194814065c5c1763db3cff3ac8d7fc260d0631f5 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 194814065c5c1763db3cff3ac8d7fc260d0631f5 -->
<!-- homu: {"type":"Approved","sha":"194814065c5c1763db3cff3ac8d7fc260d0631f5","approver":"nikomatsakis"} -->
